src/Plugins/Devices/DeviceProxy_LibUSB.cpp

```diff
@@ -454,10 +462,16 @@ void DeviceProxy_LibUSB::receive_data(uint8_t endpoint, uint8_t attributes, uint
 		break;
 	case USB_ENDPOINT_XFER_BULK:
 		*dataptr = (uint8_t *) malloc(maxPacketSize * 8);
-		rc = libusb_bulk_transfer(dev_handle, endpoint, *dataptr, maxPacketSize, length, timeout);
-		if (rc == LIBUSB_SUCCESS && debugLevel > 2)
-			cerr << "received bulk msg (" << *length << " bytes)" << endl;
-		break;
+		/* Edited by Skazza 20-10-2016 */
+		do {
+			rc = libusb_bulk_transfer(dev_handle, endpoint, *dataptr, maxPacketSize, length, timeout);
+			if (rc == LIBUSB_SUCCESS && debugLevel > 2)
+				cerr << "received bulk msg (" << *length << " bytes)" << endl;
+			if ((rc == LIBUSB_ERROR_PIPE || rc == LIBUSB_ERROR_TIMEOUT))
+				libusb_clear_halt(dev_handle, endpoint);
+			
+			attempt++;
+		} while ((rc == LIBUSB_ERROR_PIPE || rc == LIBUSB_ERROR_TIMEOUT) && attempt < MAX_ATTEMPTS);
```

Comment:
I think this change looks good, but I'm still interested in how it affects other devices, such as HIDs.